src/compute/substring.rs

```diff
@@ -24,23 +24,10 @@ use crate::{
 };
 
 fn utf8_substring<O: Offset>(array: &Utf8Array<O>, start: O, length: &Option<O>) -> Utf8Array<O> {
-    let validity = array.validity();
-    let offsets = array.offsets();
-    let values = array.values().as_slice();
-
-    let mut new_offsets = MutableBuffer::<O>::with_capacity(array.len() + 1);
-    let mut new_values = MutableBuffer::<u8>::new(); // we have no way to estimate how much this will be.
-
-    new_offsets.push(O::zero());
+    let length = length.map(|v| v.to_usize());
 
-    offsets.windows(2).for_each(|windows| {
-        // Safety:
-        // invariant of the struct that these values are utf8
-        let str_val = unsafe {
-            std::str::from_utf8_unchecked(&values[windows[0].to_usize()..windows[1].to_usize()])
-        };
-
-        // compute where we should start slicing this entry
+    let iter = array.values_iter().map(|str_val| {
```

Comment:
Oh, yes, I was already thinking that the iterators don't have bounds checks. But I wanted to avoid reading the validity buffer. But we can also only iter the values. Much better.